guppylang/prelude/_internal.py

```diff
@@ -277,22 +283,34 @@ def check(self, args: list[ast.expr], ty: Type) -> tuple[ast.expr, Subst]:
 
 
 class ResultChecker(CustomCallChecker):
-    """Call checker for the `result` function.
-
-    This is a temporary hack until we have implemented the proper results mechanism.
-    """
+    """Call checker for the `result` function."""
 
     def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
         check_num_args(2, len(args), self.node)
         [tag, value] = args
-        if not isinstance(tag, ast.Constant) or not isinstance(tag.value, int):
-            raise GuppyTypeError("Expected an int literal", tag)
+        if not isinstance(tag, ast.Constant) or not isinstance(tag.value, str):
+            raise GuppyTypeError("Expected a string literal", tag)
         value, ty = ExprSynthesizer(self.ctx).synthesize(value)
-        if ty.linear:
-            raise GuppyTypeError(
-                f"Cannot use value with linear type `{ty}` as a result", value
-            )
-        return with_loc(self.node, ResultExpr(value, ty, tag.value)), NoneType()
+        # We only allow numeric values or vectors of numeric values
+        err = (
+            f"Expression of type `{ty}` is not a valid result. Only numeric values or "
+            "arrays thereof are allowed."
+        )
+        if isinstance(ty, NumericType) or is_bool_type(ty):
+            base_ty = ty
+            array_len: Const | None = None
+        elif is_array_type(ty):
+            [ty_arg, len_arg] = ty.args
+            assert isinstance(ty_arg, TypeArg)
+            assert isinstance(len_arg, ConstArg)
```

Comment:
can these asserts result in unhelpful errors?

Reply:
No, failing asserts would be a compiler bug since you shouldn't be able to construct an array type with invalid args